devices: constify device pointers

gmarull opened the pull request:
It is frequent to find variable definitions like this:

```c
static const struct device *dev = DEVICE_DT_GET(...);
```

That is, module level variables that are statically initialized with a
device reference. Such value is, in most cases, never changed meaning
the variable can also be declared as const (immutable):

```c
static const struct device *const dev = DEVICE_DT_GET(...);
```

This PR constifies device pointers and initializes devices at compile time when possible (allowing constification).

gmarull committed:
drivers: bluetooth: hci: initialize uart device at compile time

The UART device can be initialized at compile time, allowing to constify
the device pointer. Also fix return value to -ENODEV if device is not
ready.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: clock_control: nrf: initialize temp sensor at compile time

Initialize temperature sensor at compile time, allowing to constify
device pointer.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: console: initialize uart devices at compile time

A few drivers could initialize chosen UART at compile time, allowing to
constify device pointer.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: counter: mcux_qtmr: initialize timer devices at compile time

Initialize timer devices at compile time. This allows to constify the
array of timer devices.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: gpio: npcx: remove redundant declarations

Remove redundant declarations by definining and initializing before
variables are used.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: intc: ioapic: initialize vtd device at compile time

Initialize the vtd device at compile time. Also removed unnecessary
macro hackery. As a result, vtd pointer can be constified.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: intc: miwu: remove redundant declarations

Remove redundant definitions by defining at the top before variable is
used. Also constify device array.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: serial: pipe: init device at compile time and check if ready

Initialize UART pipe device at compile time, allowing to constify the
device pointer. Also fix device ready check (was checking for NULL, an
always true condition when using DEVICE_DT_GET).

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
samples: boards: reel_board: mesh_badge: initialize epd at compile time

Initialize epd device at compile time, so we can constify epd device
pointer.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
tests: pm: initialize devices at compile time and fix ready checks

Initialize tests devices a,c at compile time. Also fix ready check (was
wrongly checking for NULL).

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
tests: drivers: uart: initialize devices at compile time

Initialize devices at compile time. Also fix some ready check problems
(e.g. in async test readiness failure just resulted in a printk, but
test continued...).

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
tests: drivers: gpio: initialize device at compile time

Initialize device at compile time, allowing to constify device pointer.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
tests: boards: mec15xxevb_assy6853: initialize device at compile time

Initialize SPI at compile time, allowing to constify the device pointer.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
samples: boards: 96b_argonkey: sensor: init device at compile time

Initialize device at compile time so that device pointer can be
constified.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
samples: boards: stm32: serial_wakeup: fix device usage issues

Device ready check was wrong (< 0 ?!), so fix it. Also drop static
qualifier for the device variable (not needed) and initialize at
definition time.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
tests: boards: intel_adsp: initialize devices at compile time

Initialize devices at compile time so that pointer can be constified.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
random: initialize devices at compile time

Initialize devices at compile time so that device pointer can be
constified.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
samples: drivers: peci: improve test

- Require peci-0 alias (test without it is pointless)
- Initialize peci device at compile time
- Filter depending on alias
- Remove redundant peci tag

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
samples: boards: stm32: usbc: initialize devices at compile time

Initialize ADC device at compile time, allowing to constify the device
pointer.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
samples: drivers: espi: initialize devices at compile time

Initialize devices at compile time so that device pointers can be
constified.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
samples: net: gsm_modem: initialize devices at compile time

Initialize devices at compile time so that device pointers can be
constified.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
samples: net: wpan_serial: initialize devices at compile time

Initialize devices at compile time so that device pointers can be
constified.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
samples: ipc: initialize devices at compile time

Initialize devices at compile time so that device pointers can be
constified.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
bluetooth: host: initialize devices as compile time

Initialize monitor UART device at compile time so that device pointer
can be constified.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
samples: usb: audio: initialize devices at compile time

Initialize devices at compile time, allowing to constify device pointer.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
fb: initialize devices at compile time

Initialize devices at compile time, allowing to constify device pointer.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
ipc: rpmsg_service: initialize devices at compile time

Initialize devices at compile time, allowing to constify device pointer.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
tracing: initialize devices at compile time

Initialize devices at compile time, allowing to constify device pointer.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
logging: initialize devices at compile time

Initialize devices at compile time, allowing to constify device pointer.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
tests: bluetooth: initialize devices at compile time

Initialize devices at compile time, allowing to constify device pointer.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
devices: constify statically initialized device pointers

It is frequent to find variable definitions like this:

```c
static const struct device *dev = DEVICE_DT_GET(...)
```

That is, module level variables that are statically initialized with a
device reference. Such value is, in most cases, never changed meaning
the variable can also be declared as const (immutable). This patch
constifies all such cases.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

jfischer-no commented on modules/mbedtls/zephyr_init.c:
```diff
@@ -45,7 +45,7 @@ static void init_heap(void)
 #endif /* CONFIG_MBEDTLS_ENABLE_HEAP && MBEDTLS_MEMORY_BUFFER_ALLOC_C */
 
-static const struct device *const entropy_dev =
+static const struct device *const const entropy_dev =
```
double `const`

gmarull replied:
fixed

jfischer-no commented on modules/openthread/platform/radio.c:
```diff
@@ -79,7 +79,7 @@ static uint8_t ack_psdu[ACK_PKT_LENGTH];
 static struct net_buf *tx_payload;
 
-static const struct device *const radio_dev =
+static const struct device *const const radio_dev =
```
double `const`

jfischer-no reviewed (commented):
TBH, I am not entirely convinced, even I always use it for function parameter where possible.

galak commented:
Should we introduce some `typedef` to handle the `const's` as I feel people aren't going to get this correct.

gmarull commented:
> Should we introduce some `typedef` to handle the `const's` as I feel people aren't going to get this correct.

something like

```c
typedef const struct device* device_t;
```

?

carlocaione commented:
> Should we introduce some `typedef`

I thought that the agreed answer to this sort of questions was always no :) (half /jk)

anangl commented on drivers/gpio/gpio_npcx.c:
```diff
@@ -18,9 +18,13 @@
+	DT_INST_FOREACH_STATUS_OKAY(NPCX_GPIO_DEV)
+};
+
+static int gpio_devs_count = ARRAY_SIZE(gpio_devs);
```
This can be also `const`, but since the value is used only in one place, it would be actually better to use `ARRAY_SIZE(gpio_devs)` directly there.

gmarull replied:
thanks, simplified by dropping gpio_devs_count

anangl commented on samples/net/gsm_modem/src/main.c:
```diff
@@ -114,8 +114,6 @@ int main(void)
 {
 	const struct device *uart_dev = DEVICE_DT_GET(UART_NODE);
```
nit: this can be also constified.

gmarull replied:
I've only tackled statically defined variables in this PR (just to scope the changes a little bit)

anangl commented on samples/net/wpan_serial/src/main.c:
```diff
@@ -56,12 +56,13 @@ static uint8_t slip_buf[1 + 2 * CONFIG_NET_BUF_DATA_SIZE];
 static struct ieee802154_radio_api *radio_api;
-static const struct device *const ieee802154_dev =
+static const struct device *ieee802154_dev =
 	DEVICE_DT_GET(DT_CHOSEN(zephyr_ieee802154));
```
Why is the `const` removed here (same applies to `samples/net/wpanusb/src/wpanusb.c` in the same commit)?

gmarull replied:
already fixed

anangl replied:
Not really. See cc5ce04dde4cd17fea71cdc78b506e3a9e9a37b0. The change is then reverted in https://github.com/zephyrproject-rtos/zephyr/pull/49167/commits/2548ce925c341b5be34a6825d67360306a5f78e8#diff-7a5b555058d53505817a540a1c8e9e5817e12132c5bd316c890552dfbc2b5170L59-R59.

gmarull replied:
I see, that looks like a rebase artifact will fix it. thanks :-)

anangl commented on subsys/tracing/tracing_backend_uart.c:
```diff
@@ -20,7 +20,8 @@
 
 
-static const struct device *tracing_uart_dev;
+static const struct device *const  =
```
Anonymous variable? :wink:

gmarull replied:
argh, fixed

gmarull committed:
samples: drivers: ps2: fix test issues

- Test effectively requires a microchip,xec-ps2, but it was "optional"
  even though the test continued setting up a callback that makes use of
  the device.
- Improve test filtering so that it requires a valid compatible to be
  enabled (allows to delete redundant ps2 tag)
- Initialize device at compile time (allowing to constify its pointer)

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull commented:
note: compliance check seems to be a false positive.
